driver/apdu/uqmi.c

```diff
@@ -0,0 +1,273 @@
+// SPDX-License-Identifier: MIT
+/* Copyright (c) 2024 David Bauer <david.bauer@uniberg.com> */
+
+#include "at.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <cjson/cJSON_ex.h>
+#include <euicc/interface.h>
+#include <euicc/hexutil.h>
+
+static FILE *fuart;
+static int client_id = 0;
+static int logic_channel = 0;
+static char *devpath = NULL;
+
+
+static int uqmi_execute_command(const char *command, char *buf, size_t bufsize)
+{
+    char *debug_str = getenv("LPAC_QMI_DEBUG");
+    char final_command[2048] = {};
+    FILE *fp;
+    char *remaining;
+    int debug = debug_str ? atoi(debug_str) : 0;
+
+    if (debug)
+        printf("UQMI_DEBUG_TX: %s\n", command);
+
+    if (snprintf(final_command, sizeof(final_command), "uqmi -s -d %s %s", devpath, command) >= sizeof(final_command))
+    {
+        fprintf(stderr, "Command too long\n");
+        return -1;
+    }
+
+    fp = popen(final_command, "r");
+    if (fp == NULL)
+    {
+        fprintf(stderr, "Failed to execute command: %s\n", command);
+        return -1;
+    }
+
+    remaining = buf;
+    if (buf) {
+        while (fgets(remaining, bufsize - (remaining - buf), fp) != NULL)
+        {
+            /* Read to buffer */
+        }
+        if (debug)
+            printf("UQMI_DEBUG_RX: %s", buf);
+    }
+
+    pclose(fp);
+    return 0;
+}
+
+static int uqmi_open_client()
+{
+    char buffer[2048] = {};
+    int ret;
+
+    ret = uqmi_execute_command("--get-client-id uim", buffer, sizeof(buffer));
+    if (ret)
+    {
+        return -1;
+    }
+
+    client_id = atoi(buffer);
+    if (client_id == 0)
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int apdu_interface_connect(struct euicc_ctx *ctx)
+{
+    const char *device;
+
+    client_id = 0;
+    logic_channel = 0;
+    devpath = getenv("LPAC_QMI_DEV");
+
+    return uqmi_open_client();
+}
+
+static void apdu_interface_disconnect(struct euicc_ctx *ctx)
+{
+    char command[64] = {};
+
+    snprintf(command, sizeof(command), "--set-client-id uim,%d --release-client-id uim", client_id);
+    uqmi_execute_command(command, NULL, 0);
+
+    client_id = 0;
+    logic_channel = 0;
+}
+
+static int apdu_interface_transmit(struct euicc_ctx *ctx, uint8_t **rx, uint32_t *rx_len, const uint8_t *tx, uint32_t tx_len)
+{
+    cJSON *root = NULL, *jtmp = NULL;
+    char cmd[2048] = {};
+    char buf[2048] = {};
+    char *json;
+
+    *rx = NULL;
+    *rx_len = 0;
+
+    if (!client_id)
+    {
+        return -1;
+    }
+
+    if (!logic_channel)
+    {
+        return -1;
+    }
+
+    for (int i = 0; i < tx_len; i++)
+    {
+        snprintf(buf + strlen(buf), sizeof(buf) - strlen(buf), "%02X", (uint8_t)(tx[i] & 0xFF));
+    }
```

Comment:
builtin bin to hex string function, plz use `euicc_hexutil_bin2hex`

https://github.com/estkme-group/lpac/blob/9e0612b68dec3f02e8fac8997d8dd653c369f42b/euicc/hexutil.h#L4-L8